fdbserver/ClusterRecovery.actor.cpp

```diff
@@ -1172,6 +1173,23 @@ ACTOR Future<Void> readTransactionSystemState(Reference<ClusterRecoveryData> sel
 	self->txnStateStore =
 	    keyValueStoreLogSystem(self->txnStateLogAdapter, self->dbgid, self->memoryLimit, false, false, true);
 
+	// Version 0 occurs at the version epoch. The version epoch can be set
+	// through the management API, otherwise a default timestamp is used. The
+	// version epoch is the number of seconds since the Unix epoch.
+	// TODO: Set `versionEpochKey` to the default value on new clusters to
+	// enable new clusters to use a version in-step with wall clock time.
+	self->versionEpoch.reset();
+	Optional<Standalone<StringRef>> versionEpochValue = wait(self->txnStateStore->readValue(versionEpochKey));
+	self->versionEpoch = SERVER_KNOBS->DEFAULT_VERSION_EPOCH;
+	if (versionEpochValue.present()) {
+		self->versionEpoch = BinaryReader::fromStringRef<Version>(versionEpochValue.get(), Unversioned());
+	} else if (BUGGIFY) {
+		// Cannot use a positive version epoch in simulation because of the
+		// clock starting at 0. A positive version epoch would mean the initial
+		// cluster version was negative.
+		self->versionEpoch = deterministicRandom()->randomInt64(-1e6, 0);
```

Comment:
@sfc-gh-ajbeamon this is how the version epoch is set randomly in simulation. Playing around with this value a bit, it may have an effect on whether simulation passes. If it is too large, restart tests may not work? I'm not totally sure yet.